### PR TITLE
vision model error analysis

### LINKED ISSUE #915: [FEAT][CV] Vision model error analysis
In the tabular package, the model error analysis check automatically detects segments in which the model error is high. In the vision package, this can be done based on the image properties. Otherwise, check should be pretty much the same.

### PR DESCRIPTION
**Pull request by benisraeldan**
#### Reference Issues/PRs

Resolves #915 

#### What does this implement/fix? Explain your changes.

Refactor model error analysis to work both on tabular and vision, and add vision specific logic

### CODE REVIEW

**Review by noamzbr — COMMENTED** [review]
There are unnecessary changes in the class_performance notebok.

Other then that - mainly organizing the code is needed

File: deepchecks/vision/checks/performance/model_error_analysis.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+                                                                        test_property_df,
+                                                                        self.test_scores,
+                                                                        train_property_df.columns.to_list(),
+                                                                        [],
+                                                                        min_error_model_score=self.min_error_model_score,
+                                                                        random_state=self.random_state)
+
+        display, value = error_model_display(error_fi,
+                                      test_property_df,
+                                      DeepcheckScorer('accuracy', 'Accuracy'),
```
[the remark this answers is not in the document]
Comment (benisraeldan) [reply deepchecks/vision/checks/performance/model_error_analysis.py:175]: It's unused in vision, I'll change it to null and reorder the parameters, we use the predictions as metrics.

File: deepchecks/vision/checks/performance/model_error_analysis.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+                    getattr(dataset.image_formatter, image_property)(images)
+                )
+            elif callable(image_property):
+                properties[image_property.__name__].extend(image_property(images))
+            else:
+                raise DeepchecksValueError(
+                    'Do not know how to work with image'
+                    f'property of type - {type(image_property).__name__}'
+                )
+
```
Comment (ItayGabbay) [inline deepchecks/vision/checks/performance/model_error_analysis.py:131]: Maybe better to detach to CPU here

File: deepchecks/vision/checks/performance/model_error_analysis.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+                        continue
+
+                    ious = compute_class_ious(detected, ground_truth)
+                    count = 0
+                    sum_iou = 0
+                    for _, cls_ious in ious.items():
+                        for detection in cls_ious:
+                            if len(detection):
+                                sum_iou += max(detection)
+                                count += 1
```
Comment (ItayGabbay) [inline deepchecks/vision/checks/performance/model_error_analysis.py:154]: what happens if len(detection)==0?

File: deepchecks/vision/checks/performance/model_error_analysis.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+                    if count:
+                        mean_ious.append(sum_iou/count)
+                    else:
+                        mean_ious.append(0)
+
+                return mean_ious
+
```
Comment (ItayGabbay) [inline deepchecks/vision/checks/performance/model_error_analysis.py:161]: Maybe it's better to define the loss functions outside of the `update`?

File: deepchecks/utils/performance/error_model.py [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+    min_feature_contribution : float
+        Minimum value to consider a feature to output.
+    n_display_samples : int
+        Maximum number of values to represent in the display
+    min_segment_size : int
+        Minimum segment size to consider.
+    random_state: int
+        Random seed
+    """
+    print(n_display_samples)
```
Comment (ItayGabbay) [inline deepchecks/utils/performance/error_model.py:151]: Delete this

File: deepchecks/utils/performance/error_model.py [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+        Maximum number of features to output.
+    min_feature_contribution : float
+        Minimum value to consider a feature to output.
+    n_display_samples : int
+        Maximum number of values to represent in the display
+    min_segment_size : int
+        Minimum segment size to consider.
+    random_state: int
+        Random seed
+    """
```
Comment (ItayGabbay) [inline deepchecks/utils/performance/error_model.py:150]: Missing Returns

**Comment by benisraeldan** [comment]
Color are inverted, because higher score is worse than lower score in these cases, also there seem to be an error in the calculation of binary cross entropy

File: deepchecks/utils/performance/error_model.py [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+        ]
+    )
+
+    return Pipeline(steps=[
+        ('preprocessing', preprocessor),
+        ('model', RandomForestRegressor(max_depth=4, n_jobs=-1, random_state=random_state))
+    ]), numeric_features + cat_features
+
+
+def error_model_display_dataframe(error_fi: pd.Series,
```
Comment (matanper) [inline deepchecks/utils/performance/error_model.py:85]: not really clear for me why you need this function?

File: deepchecks/utils/single_sample_metrics.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+# Deepchecks is distributed under the terms of the GNU Affero General
+# Public License (version 3 or later).
+# You should have received a copy of the GNU Affero General Public License
+# along with Deepchecks.  If not, see <http://www.gnu.org/licenses/>.
+# ----------------------------------------------------------------------------
+#
+"""Common metrics to calculate performance on single samples."""
+import numpy as np
+
+from deepchecks.vision.metrics_utils.iou_utils import compute_class_ious
```
Comment (matanper) [inline deepchecks/utils/single_sample_metrics.py:14]: I think we can't import vision from deepchecks base

File: deepchecks/vision/metrics_utils/iou_utils.py [hunk 9]
```diff
@@ -45,3 +48,24 @@ def compute_pairwise_ious(detected, ground_truth):
...
             ious[d_idx, g_idx] = jaccard_iou(d, g)
     return ious
+
+
+def compute_class_ious(detected, ground_truth):
+    """Compute ious between bounding boxes of the same class."""
+    bb_info = defaultdict(lambda: {"detected": [], "ground_truth": []})
+
+    for d in detected:
+        if isinstance(d[5], torch.Tensor):
```
Comment (matanper) [inline deepchecks/vision/metrics_utils/iou_utils.py:58]: we have convention that predictions are always tensor no?

File: deepchecks/vision/metrics_utils/iou_utils.py [hunk 10]
```diff
@@ -45,3 +48,24 @@ def compute_pairwise_ious(detected, ground_truth):
         for d_idx, d in enumerate(detected):
             ious[d_idx, g_idx] = jaccard_iou(d, g)
     return ious
+
+
+def compute_class_ious(detected, ground_truth):
```
Comment (matanper) [inline deepchecks/vision/metrics_utils/iou_utils.py:53]: why not use the metric class?

Comment (benisraeldan) [reply deepchecks/vision/metrics_utils/iou_utils.py:53]: metric class aggregates on batches, and returns a response on all the data, so it doesn't make sense to use it here

**Review by matanper — APPROVED** [review]
